airflow/providers/facebook/ads/hooks/ads.py

```diff
@@ -118,11 +122,32 @@ def bulk_facebook_report(
         :param sleep_time: Time to sleep when async call is happening
         :type sleep_time: int
 
-        :return: Facebook Ads API response, converted to Facebook Ads Row objects
-        :rtype: List[AdsInsights]
+        :return: Facebook Ads API response,
+            converted to Facebook Ads Row objects regarding given Account ID type
+        :rtype: List[AdsInsights] or Dict[str, List[AdsInsights]]
         """
         api = self._get_service()
-        ad_account = AdAccount(api.get_default_account_id(), api=api)
+        if self.is_backward:
+            return self._facebook_report(
+                account_id=self.facebook_ads_config["account_id"],
+                api=api,
+                params=params,
+                fields=fields,
+                sleep_time=sleep_time,
+            )
```

Comment:
This logic does not look… backward to me. Why is the flag called `is_backward` in the first place? What does backward mean in this context?

Reply:
Hi @uranusjr, In this context, backward means it is returning the same as the old version which is List[AdsInsights]. In the current version, Facebook Ads Hook only supports one account_id (string) at the time meaning if you have multiple account_id (array|list) for ads, you have to maintain multiple Facebook Ads Connection in the Airflow.  To support multiple account_id in the same Facebook Ads Connection, I corollated both account_id and the list of AdsInsights with returning Dict[str, List[AdsInsights]] but this was a breaking change since the return type was changed. After that, I updated the logic with this check if the account_id in the provided Facebook Connection as a string (single account_id), it returns the same format which is List[AdsInsights] and in the operator this situation also handled. If the multiple account_id is provided in the Facebook Ads Connection as an array, then the check provide to return Dict[str, List[AdsInsights]] which every report correlated with account_id to export them regarding account_id. For that reason, I called that check as is_backward which provides the check of the old connection format to work as it is.

Reply:
The implementation still looks wrong to me. The main issue is that `_get_service` should not _mutate` the hook object.

A better implementation would tie the backward compatibility switch directly against `facebook_ads_config`, instead of relying on calling `_get_service` to modify the flag. One way to achieve this is to make the flag a `cached_property` like `facebook_ads_config`:

```python
@cached_property
def multiple_accounts(self) -> bool:
    return isinstance(self.config["account_id"], list)
```

`is_backwards` is also a very bad name because it is unclear what “backwards” means in the context. It would be better to directly describe what the difference is instead. The “new” behaviour allows using multiple account, while the existing one allows only one account, so the property can be called `multiple_accounts` to signify the difference. (Or it can be called `single_account` with the semantics flipped.)

Reply:
I have changed the controlling flag to a `cached_property`. You are right. It is definitely more readable and safer to use `cached_property` and not mutate any hook object in `_get_service` in this way.